backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -365,4 +418,362 @@ public FutureExperimentRunDAO(Executor executor, FutureJdbi jdbi, UAC uac) {
+  }
+
+  private InternalFuture<ExperimentRun> insertExperimentRun(ExperimentRun newExperimentRun) {
+    final var now = Calendar.getInstance().getTimeInMillis();
+    return jdbi.withHandle(
+            handle -> {
+              Map<String, Object> runValueMap = new LinkedHashMap<>();
+              runValueMap.put("id", newExperimentRun.getId());
+              runValueMap.put("project_id", newExperimentRun.getProjectId());
+              runValueMap.put("experiment_id", newExperimentRun.getExperimentId());
+              runValueMap.put("name", newExperimentRun.getName());
+              runValueMap.put("description", newExperimentRun.getDescription());
+              runValueMap.put("date_created", newExperimentRun.getDateCreated());
+              runValueMap.put("date_updated", newExperimentRun.getDateUpdated());
+              runValueMap.put("start_time", newExperimentRun.getStartTime());
+              runValueMap.put("end_time", newExperimentRun.getEndTime());
+              runValueMap.put("code_version", newExperimentRun.getCodeVersion());
+              // TODO: code version snapshot
+              /*runValueMap.put("code_version_snapshot_id", newExperimentRun.getCodeVersionSnapshot());*/
+              runValueMap.put("job_id", newExperimentRun.getJobId());
+              runValueMap.put("parent_id", newExperimentRun.getParentId());
+              runValueMap.put("owner", newExperimentRun.getOwner());
+
+              EnvironmentBlob environmentBlob =
+                  sortPythonEnvironmentBlob(newExperimentRun.getEnvironment());
+              runValueMap.put(
+                  "environment", ModelDBUtils.getStringFromProtoObject(environmentBlob));
+              runValueMap.put("deleted", false);
+
+              String[] fieldsArr = runValueMap.keySet().toArray(new String[0]);
+              String commaFields = String.join(",", fieldsArr);
+
+              StringBuilder queryStrBuilder =
+                  new StringBuilder("insert into experiment_run ( ")
+                      .append(commaFields)
+                      .append(") values (");
+
+              for (int i = 0; i < fieldsArr.length; i++) {
+                queryStrBuilder.append(":").append(fieldsArr[i]);
+                if (i < fieldsArr.length - 1) {
+                  queryStrBuilder.append(",");
+                }
+              }
+              queryStrBuilder.append(" ) ");
+
+              LOGGER.trace("insert experiment run query string: " + queryStrBuilder.toString());
+              var query = handle.createUpdate(queryStrBuilder.toString());
+              for (Map.Entry<String, Object> objectEntry : runValueMap.entrySet()) {
+                query.bind(objectEntry.getKey(), objectEntry.getValue());
+              }
+              query.execute();
+
+              return newExperimentRun;
+            })
+        .thenCompose(
+            handle -> tagsHandler.addTags(newExperimentRun.getId(), newExperimentRun.getTagsList()),
```

Comment:
`handle` will already have been closed. Plus it's not used. Plus we're not putting this in a transaction, which is buggy since we don't have a `created` flag for ERs.

Reply:
Code is modify with better solution